contracts/account/utils/draft-ERC7579Utils.sol

```diff
@@ -169,12 +172,34 @@ library ERC7579Utils {
     }
 
     /// @dev Decodes a batch of executions. See {encodeBatch}.
+    ///
+    /// Note: This function runs some checks and will throw a {ERC7579DecodingError} if the input is not properly formatted.
     function decodeBatch(bytes calldata executionCalldata) internal pure returns (Execution[] calldata executionBatch) {
-        assembly ("memory-safe") {
-            let ptr := add(executionCalldata.offset, calldataload(executionCalldata.offset))
-            // Extract the ERC7579 Executions
-            executionBatch.offset := add(ptr, 32)
-            executionBatch.length := calldataload(ptr)
+        unchecked {
+            uint256 bufferLength = executionCalldata.length;
+
+            // Check executionCalldata is not empty
+            if (bufferLength < 32) revert ERC7579DecodingError();
+
+            // Get the offset of the array, and check its valid
+            uint256 offset = uint256(bytes32(executionCalldata[0:32]));
+
+            // Array length is between offset and offset + 32. We check that this is in the buffer
+            // Since we know executionCalldata is at least 32, we can subtract with no overflow risk.
+            if (offset > bufferLength - 32) revert ERC7579DecodingError();
+
+            // Get the array length
+            uint256 arrayLength = uint256(bytes32(executionCalldata[offset:offset + 32]));
+            if (arrayLength > type(uint64).max) revert ERC7579DecodingError();
+
+            // Get the array as a bytes slice, and check its is long enough
+            bytes calldata executionArray = executionCalldata[offset + 32:];
+            if (executionArray.length < arrayLength * 96) revert ERC7579DecodingError();
+
+            assembly ("memory-safe") {
+                executionBatch.offset := executionArray.offset
+                executionBatch.length := arrayLength
+            }
```

Comment:
Wondering if this can go into a "calldata decoding" library or similar. It's essentially a replacement for `abi.decode`

Reply:
It would need to be generic on the return type so it's not possible to make a general helper.

The safety checks could be put in a helper, but it would need to be parameterized by the size of the array elements (actually the array element [heads](https://docs.soliditylang.org/en/latest/abi-spec.html#formal-specification-of-the-encoding)) so it's not possible to make it automatically fully safe.